### PR TITLE
[SYCL] Add intel::kernel_args_restrict attribute

### PR DESCRIPTION
**Pull request by asavonic**
When applied to a device function that is invoked as a device kernel,
the attribute is a hint to the compiler that no pointer argument to
the kernel which is defined through an accessor (not USM), will alias
any other pointer kernel argument that was defined through an
accessor. This effect is equivalent to annotating restrict on all
kernel pointer arguments in an OpenCL or SPIR-V kernel.

Signed-off-by: Andrew Savonichev <andrew.savonichev@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -425,6 +427,18 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
         Attrs.insert(A);
       else if (auto *A = FD->getAttr<ReqdWorkGroupSizeAttr>())
         Attrs.insert(A);
+      else if (auto *A = FD->getAttr<SYCLIntelKernelArgsRestrictAttr>()) {
+        // Allow the intel::kernel_args_restrict only on the lambda (functor
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp]: Please do not use "functor" word.
https://github.com/intel/llvm/pull/654#issuecomment-533545781

File: clang/test/SemaSYCL/intel-restrict.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// RUN: %clang %s -fsyntax-only --sycl -DCHECKDIAG -Xclang -verify
+// RUN: %clang %s -fsyntax-only -Xclang -ast-dump --sycl | FileCheck %s
+
+[[intel::kernel_args_restrict]] // expected-warning{{'kernel_args_restrict' attribute ignored}}
+void func_ignore() {}
```
Comment (Fznamznon) [inline clang/test/SemaSYCL/intel-restrict.cpp]: Could you please check that diagnostic emitted when the attribute is applied not to function, i.e. variable, function parameter etc...

File: clang/include/clang/Basic/AttrDocs.td [hunk 3]
```diff
@@ -1835,6 +1835,43 @@ loads).
...
+capture, or functor member, of the callable to which the attribute was
+applied. This effect is equivalent to annotating restrict on **all** kernel
+pointer arguments in an OpenCL or SPIR-V kernel.
+
+If ``intel::kernel_args_restrict`` is applied to a function called from a device
+kernel, propagation of the attribute to any caller(s), including up to a kernel
+boundary, is implementation defined and not guaranteed through this
+extension. The attribute forms an unchecked assertion, in that implementations
+do not need to check/confirm the pre-condition in any way. If a user applies
+``inte::_kernel_args_restrict`` to a kernel, but there is in fact aliasing
```
[the remark this answers is not in the document]
Comment (asavonic) [reply clang/include/clang/Basic/AttrDocs.td]: Done

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (asavonic) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

**Pull request merged**